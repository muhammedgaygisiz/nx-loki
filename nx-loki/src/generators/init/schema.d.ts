/* eslint-disable */
export interface InitGeneratorSchema {}
