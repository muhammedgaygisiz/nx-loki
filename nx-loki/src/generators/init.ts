import {
  addDependenciesToPackageJson,
  ExpandedPluginConfiguration,
  formatFiles,
  NxJsonConfiguration,
  readNxJson,
  Tree,
  updateNxJson,
} from "@nx/devkit";
import { InitGeneratorSchema } from "./schema";
import { readPackageJson } from "nx/src/project-graph/file-utils";
import { LOKI_VERSION } from "../utils/versions";

const isPlugin = (plugin: string | ExpandedPluginConfiguration<unknown>) =>
  typeof plugin === "string"
    ? plugin === "nx-loki"
    : plugin.plugin === "nx-loki";

const nxLokiPlugin = {
  plugin: "nx-loki",
  options: {},
};

const attachNxLokiPlugin = (nxJson: NxJsonConfiguration<"*" | string[]>) => [
  ...nxJson.plugins,
  nxLokiPlugin,
];

const updateNxJsonWithPlugin = async (
  nxJson: NxJsonConfiguration<"*" | string[]>,
  tree: Tree
) => {
  if (!nxJson.plugins) {
    nxJson.plugins = [];
  }

  nxJson.plugins = attachNxLokiPlugin(nxJson);

  updateNxJson(tree, nxJson);
  await formatFiles(tree);
};

const addDependency = async (tree: Tree) => {
  return addDependenciesToPackageJson(
    tree,
    {},
    {
      loki: LOKI_VERSION,
    }
  );
};

export async function initGenerator(tree: Tree, options: InitGeneratorSchema) {
  const nxJson = readNxJson(tree) || {};
  const hasPlugin = nxJson.plugins?.some((plugin) => isPlugin(plugin));

  if (!hasPlugin) {
    await updateNxJsonWithPlugin(nxJson, tree);
  }

  await addDependency(tree);
}

export default initGenerator;
