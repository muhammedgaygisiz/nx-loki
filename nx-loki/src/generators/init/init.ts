import {
  addDependenciesToPackageJson,
  formatFiles,
  NxJsonConfiguration,
  readNxJson,
  Tree,
  updateNxJson,
} from "@nx/devkit";
import { InitGeneratorSchema } from "./schema";
import { LOKI_VERSION } from "../../utils/versions";
import { isLokiPlugin } from "../../utils/plugins.utils";

const nxLokiPlugin = {
  plugin: "nx-loki",
  options: {},
};

const attachNxLokiPlugin = (nxJson: NxJsonConfiguration<"*" | string[]>) => [
  ...nxJson.plugins,
  nxLokiPlugin,
];

const updateNxJsonWithPlugin = async (
  nxJson: NxJsonConfiguration<"*" | string[]>,
  tree: Tree
) => {
  if (!nxJson.plugins) {
    nxJson.plugins = [];
  }

  nxJson.plugins = attachNxLokiPlugin(nxJson);

  updateNxJson(tree, nxJson);
  await formatFiles(tree);
};

const addDependency = async (tree: Tree) => {
  return addDependenciesToPackageJson(
    tree,
    {},
    {
      loki: LOKI_VERSION,
    }
  );
};

export async function initGenerator(tree: Tree, options: InitGeneratorSchema) {
  const nxJson = readNxJson(tree) || {};
  const hasPlugin = nxJson.plugins?.some((plugin) => isLokiPlugin(plugin));

  if (!hasPlugin) {
    await updateNxJsonWithPlugin(nxJson, tree);
  }

  await addDependency(tree);
}

export default initGenerator;
