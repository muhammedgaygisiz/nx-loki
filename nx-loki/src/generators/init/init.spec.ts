import { createTreeWithEmptyWorkspace } from "@nx/devkit/testing";
import { readNxJson, Tree } from "@nx/devkit";

import { initGenerator } from "./init";
import { InitGeneratorSchema } from "./schema";

describe("init generator", () => {
  let tree: Tree;
  const options: InitGeneratorSchema = {};

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it("should run successfully", async () => {
    await initGenerator(tree, options);

    const nxJson = readNxJson(tree);
    expect(nxJson.plugins).toMatchSnapshot();

    const modifiedPackageJson = JSON.parse(tree.read("package.json", "utf-8"));

    expect(modifiedPackageJson).toMatchSnapshot();
  });
});
