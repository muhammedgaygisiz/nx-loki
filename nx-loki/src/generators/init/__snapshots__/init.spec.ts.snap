// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init generator should run successfully 1`] = `
[
  {
    "options": {},
    "plugin": "nx-loki",
  },
]
`;

exports[`init generator should run successfully 2`] = `
{
  "dependencies": {},
  "devDependencies": {
    "loki": "0.35.1",
  },
  "name": "@proj/source",
}
`;
